### DISCUSSION TITLE
为什么我是用Attu查询结果与使用Java SDK查询结果不一致。（Attu查询的为正确结果）

### DISCUSSION THREAD

**Discussion by LiJie20190102** [milvus-io/milvus#23424] [Q&A and General discussion] [answered]
背景：这个collection是使用Attu创建并且界面导入csv文件数据产生的。

使用Attu查询：
![image](https://user-images.githubusercontent.com/53458004/231916403-fd1cab7d-d283-4b59-8c3c-4b0c305b5523.png)

使用Java SDK查询：
 代码：
![image](https://user-images.githubusercontent.com/53458004/231916806-2c795aaa-dd19-4d56-8436-c2d8ef453d91.png)
结果：
![image](https://user-images.githubusercontent.com/53458004/231917155-aaab583a-88aa-42fe-bb58-746b48b1fe2f.png)

**疑问**：
1、对于java sdk而言，未出现book_id为44的记录，是不是有些参数我没设置好，所以导致结果不一样？
2、对于Attu,是根据distance来判断的排名么？是不是对应java SDK返回结果中的Score参数？
3、nprobe、Round Decimals、offset分别是什么意思呢？

**Comment by yhmo** [comment, marked as answer]
IVF是基于k-means聚类算法的索引，你可以去搜一下k-means 的原理。大致意思就是通过迭代运算把大量向量聚类成多个桶，每个桶里有若干向量。桶的数量由nlist参数指定。而搜索时的参数nprobe指定了打算在最近的几个桶里查询，根据每个桶的中心点和这条向量的距离来选取一个合适的桶来查询。这些都是向量近似近邻搜索的最基本原理。
你设置nprobe为1，它只查询一个桶，但有可能好几个桶的中心点跟这条向量的距离相等，所以有可能每次查询会打到不同的桶，所以结果有可能会变化。
你把nprobe改为5或者10再观察结果。
你去了解k-mean算法之后就很容易理解这些。round decimals, offset在文档里都有说明：https://milvus.io/api-reference/java/v2.2.5/Query%20and%20Search/search().md

round decimals是控制返回结果里的距离保留到小数点后几位。
offset是用来做pagenation，也就是结果集分页。如果返回的结果有几万条，那你不想一次拿完，就考虑分批获取。

**Reply by LiJie20190102** [reply]
1、我刚才使用attu与java SDK时，我都设置了nprobe为10，但是java SDK的查询结果中未出现正确的结果（既book_id为44的记录）
![image](https://user-images.githubusercontent.com/53458004/231958714-23d72fad-9622-48ee-9a0d-dc721d8d35cc.png)
![image](https://user-images.githubusercontent.com/53458004/231958825-eee1c104-57fd-4b39-bfbd-be1a28dd6c9c.png)

**Reply by yhmo** [reply]
把那个SEARCH_PARAM里的“offset: 5”去掉吧，不然它给你的结果集是从第5个开始给你。

**Reply by LiJie20190102** [reply]
谢谢了，不好意思，这个offset之前理解有误

**Reply by LiJie20190102** [reply]
谢谢了，不好意思，这个offset之前理解有误
